app/src/main/java/com/woefe/shoppinglist/shoppinglist/ShoppingList.java

```diff
@@ -195,24 +203,28 @@ public void sort(Comparator<? super ListItem> c) {
         notifyListChanged(Event.newOther());
     }
 
-    public void setChecked(int index, boolean isChecked) {
-        get(index).setChecked(isChecked);
-        notifyListChanged(Event.newItemChanged(index));
+    public void setChecked(ListItem item, boolean isChecked, int absolutePosition) {
+        get(indexOf(item)).setChecked(isChecked);
+        notifyListChanged(Event.newItemChanged(absolutePosition));
     }
 
-    public void toggleChecked(int index) {
-        setChecked(index, !get(index).isChecked());
+    public void toggleChecked(ListItem item, int absolutePosition) {
+        setChecked(item, !item.isChecked(), absolutePosition);
     }
 
-    public void move(int oldIndex, int newIndex) {
+    public void move(ListItem from, ListItem to) {
+        int oldIndex = indexOf(from);
+        int newIndex = indexOf(to);
```

Comment:
This could be problematic. What if an item has been added twice? The GUI probably creates new objects for every item, though. What was the rationale for this change?

Reply:
This addresses the problem you described above: GUI is disconnected from flat last.
Instead of relying on indexes, this is now really using the objects as stored in ArrayList.
Therefore it uses the real object and moves them in the List.
(upon write it then just goes through all ordered categories and writes the items to list.